File: api/routers/goals.py
import logging
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException
from api.services.goal_service import GoalService
from api.dependencies import get_current_user_id
from api.schemas.goals import GoalCreate, GoalUpdate

logger = logging.getLogger(__name__)


def create_goal_router(goal_service: GoalService) -> APIRouter:
    router = APIRouter()

    @router.get("/goals")
    def list_goals(user_id: int = Depends(get_current_user_id)):
        try:
            return goal_service.list_goals(user_id)
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    @router.post("/goals", status_code=201)
    def create_goal(body: GoalCreate, user_id: int = Depends(get_current_user_id)):
        try:
            title = body.title.strip()
            if not title:
                raise HTTPException(status_code=400, detail="Title is required")
            description = (body.description or "").strip()
            goal_id = goal_service.create_goal(user_id, title, description, body.frequency_per_week)
            return {"id": goal_id}
        except HTTPException:
            raise
        except ValueError as e:
            raise HTTPException(status_code=400, detail=str(e))
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    @router.get("/goals/{goal_id}")
    def get_goal(goal_id: int, user_id: int = Depends(get_current_user_id)):
        try:
            goal = goal_service.get_goal(user_id, goal_id)
            if not goal:
                raise HTTPException(status_code=404, detail="Not found")
            return goal
        except HTTPException:
            raise
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    @router.put("/goals/{goal_id}")
    @router.patch("/goals/{goal_id}")
    def update_goal(goal_id: int, body: GoalUpdate, user_id: int = Depends(get_current_user_id)):
        try:
            frequency = body.frequency_per_week if body.frequency_per_week is not None else body.frequency
            success = goal_service.update_goal(
                user_id,
                goal_id,
                body.title,
                body.description,
                int(frequency) if frequency is not None else None,
            )
            if not success:
                raise HTTPException(status_code=404, detail="No changes or goal not found")
            return {"status": "ok"}
        except HTTPException:
            raise
        except ValueError as e:
            raise HTTPException(status_code=400, detail=str(e))
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    @router.delete("/goals/{goal_id}")
    def delete_goal(goal_id: int, user_id: int = Depends(get_current_user_id)):
        try:
            success = goal_service.delete_goal(user_id, goal_id)
            if not success:
                raise HTTPException(status_code=404, detail="Not found")
            return {"status": "ok"}
        except HTTPException:
            raise
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    @router.post("/goals/{goal_id}/progress")
    def increment_progress(goal_id: int, user_id: int = Depends(get_current_user_id)):
        try:
            updated = goal_service.increment_progress(user_id, goal_id)
            if not updated:
                raise HTTPException(status_code=404, detail="Not found")
            return updated
        except HTTPException:
            raise
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    @router.get("/goals/{goal_id}/completions")
    @router.get("/goal/{goal_id}/completions")
    def get_completions(
        goal_id: int,
        start: Optional[str] = None,
        end: Optional[str] = None,
        user_id: int = Depends(get_current_user_id),
    ):
        try:
            return goal_service.get_completions(user_id, goal_id, start, end)
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    return router

File: functions/_shared/routes/goals.ts
import { Hono } from 'hono';
import { zValidator } from '@hono/zod-validator';
import { z } from 'zod';
import { HTTPException } from 'hono/http-exception';
import type { AppEnv } from '../lib/env';
import { createDb } from '../db/client';
import { goalCreateSchema, goalUpdateSchema } from '@shared/schemas/goals';
import {
  ValidationError,
  createGoal,
  deleteGoal,
  getGoalById,
  getGoalCompletions,
  getGoals,
  incrementGoalProgress,
  updateGoal,
} from '../services/goal-service';

const goalIdParamSchema = z.object({ goal_id: z.coerce.number().int() });
const completionsQuerySchema = z.object({ start: z.string().optional(), end: z.string().optional() });

export const goalRoutes = new Hono<AppEnv>();

goalRoutes.get('/goals', async (c) => {
  const userId = c.get('userId');
  const db = createDb(c.env.DB);
  return c.json(await getGoals(db, userId));
});

goalRoutes.post('/goals', zValidator('json', goalCreateSchema), async (c) => {
  const userId = c.get('userId');
  const body = c.req.valid('json');
  const db = createDb(c.env.DB);

  const title = body.title.trim();
  if (!title) {
    throw new HTTPException(400, { message: 'Title is required' });
  }

  try {
    const goalId = await createGoal(db, userId, { ...body, title });
    return c.json({ id: goalId }, 201);
  } catch (err) {
    if (err instanceof ValidationError) {
      throw new HTTPException(400, { message: err.message });
    }
    throw err;
  }
});

goalRoutes.get('/goals/:goal_id', zValidator('param', goalIdParamSchema), async (c) => {
  const userId = c.get('userId');
  const { goal_id: goalId } = c.req.valid('param');
  const db = createDb(c.env.DB);

  const goal = await getGoalById(db, userId, goalId);
  if (!goal) {
    throw new HTTPException(404, { message: 'Not found' });
  }
  return c.json(goal);
});

async function handleUpdateGoal(
  db: ReturnType<typeof createDb>,
  userId: number,
  goalId: number,
  body: z.infer<typeof goalUpdateSchema>
) {
  const frequency = body.frequency_per_week ?? body.frequency;

  try {
    const success = await updateGoal(db, userId, goalId, { ...body, frequency_per_week: frequency });
    if (!success) {
      throw new HTTPException(404, { message: 'No changes or goal not found' });
    }
    return { status: 'ok' as const };
  } catch (err) {
    if (err instanceof HTTPException) throw err;
    if (err instanceof ValidationError) {
      throw new HTTPException(400, { message: err.message });
    }
    throw err;
  }
}

goalRoutes.put(
  '/goals/:goal_id',
  zValidator('param', goalIdParamSchema),
  zValidator('json', goalUpdateSchema),
  async (c) => {
    const userId = c.get('userId');
    const { goal_id: goalId } = c.req.valid('param');
    const body = c.req.valid('json');
    const db = createDb(c.env.DB);
    return c.json(await handleUpdateGoal(db, userId, goalId, body));
  }
);
goalRoutes.patch(
  '/goals/:goal_id',
  zValidator('param', goalIdParamSchema),
  zValidator('json', goalUpdateSchema),
  async (c) => {
    const userId = c.get('userId');
    const { goal_id: goalId } = c.req.valid('param');
    const body = c.req.valid('json');
    const db = createDb(c.env.DB);
    return c.json(await handleUpdateGoal(db, userId, goalId, body));
  }
);

goalRoutes.delete('/goals/:goal_id', zValidator('param', goalIdParamSchema), async (c) => {
  const userId = c.get('userId');
  const { goal_id: goalId } = c.req.valid('param');
  const db = createDb(c.env.DB);

  const success = await deleteGoal(db, userId, goalId);
  if (!success) {
    throw new HTTPException(404, { message: 'Not found' });
  }
  return c.json({ status: 'ok' });
});

goalRoutes.post('/goals/:goal_id/progress', zValidator('param', goalIdParamSchema), async (c) => {
  const userId = c.get('userId');
  const { goal_id: goalId } = c.req.valid('param');
  const db = createDb(c.env.DB);

  const updated = await incrementGoalProgress(db, userId, goalId);
  if (!updated) {
    throw new HTTPException(404, { message: 'Not found' });
  }
  return c.json(updated);
});

goalRoutes.get(
  '/goals/:goal_id/completions',
  zValidator('param', goalIdParamSchema),
  zValidator('query', completionsQuerySchema),
  async (c) => {
    const userId = c.get('userId');
    const { goal_id: goalId } = c.req.valid('param');
    const { start, end } = c.req.valid('query');
    const db = createDb(c.env.DB);
    return c.json(await getGoalCompletions(db, userId, goalId, start, end));
  }
);
goalRoutes.get(
  '/goal/:goal_id/completions',
  zValidator('param', goalIdParamSchema),
  zValidator('query', completionsQuerySchema),
  async (c) => {
    const userId = c.get('userId');
    const { goal_id: goalId } = c.req.valid('param');
    const { start, end } = c.req.valid('query');
    const db = createDb(c.env.DB);
    return c.json(await getGoalCompletions(db, userId, goalId, start, end));
  }
);
